File: android/src/main/kotlin/com/appodeal/appodeal_flutter/AppodealNative.kt
package com.appodeal.appodeal_flutter

import com.appodeal.ads.NativeAd
import com.appodeal.ads.NativeCallbacks
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

internal class AppodealNative(
    private val flutterPluginBinding: FlutterPlugin.FlutterPluginBinding,
) : MethodChannel.MethodCallHandler {

    val adChannel: MethodChannel by lazy {
        MethodChannel(
            flutterPluginBinding.binaryMessenger,
            "appodeal_flutter/native"
        ).apply { setMethodCallHandler(this@AppodealNative) }
    }

    val adListener: Listener = Listener(adChannel)

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {}

    internal class Listener(private val adChannel: MethodChannel) : NativeCallbacks {

        override fun onNativeLoaded() {
            adChannel.invokeMethod("onNativeLoaded", null)
        }

        override fun onNativeFailedToLoad() {
            adChannel.invokeMethod("onNativeFailedToLoad", null)
        }

        override fun onNativeShown(nativeAd: NativeAd?) {
            adChannel.invokeMethod("onNativeShown", null)
        }

        override fun onNativeShowFailed(nativeAd: NativeAd?) {
            adChannel.invokeMethod("onNativeShowFailed", null)
        }

        override fun onNativeClicked(nativeAd: NativeAd?) {
            adChannel.invokeMethod("onNativeClicked", null)
        }

        override fun onNativeExpired() {
            adChannel.invokeMethod("onNativeExpired", null)
        }
    }
}

File: ios/Classes/AppodealNative.swift
import Flutter
import Foundation
import Appodeal

internal final class AppodealNative {
    
    internal let adChannel: FlutterMethodChannel
    internal let adListener: Listener
    
    private lazy var settings: APDNativeAdSettings = {
        let adSettings = APDNativeAdSettings.default()
        //        adSettings.adViewClass = NativeAdView.self
        //        adSettings.autocacheMask = [.icon, .media]
        //        adSettings.type = .video
        return adSettings
    }()
    
    private lazy var nativeAdQueue: APDNativeAdQueue = {
        APDNativeAdQueue(
            sdk: nil,
            settings: settings,
            delegate: adListener,
            autocache: true
        )
    }()
    
    private lazy var nativeArray: [APDNativeAd] = []
    
    init(registrar: FlutterPluginRegistrar) {
        adChannel = FlutterMethodChannel(name: "appodeal_flutter/native", binaryMessenger: registrar.messenger())
        adListener = Listener(adChannel: adChannel)
    }
    
    func load() {
        nativeAdQueue.setMaxAdSize(1)
        nativeAdQueue.loadAd()
    }
    
    func getNativeAd() -> APDNativeAd? {
        let nativeAds = nativeAdQueue.getNativeAds(ofCount: 1)
        let nativeAd = nativeAds.isEmpty ? nil : nativeAds.first
        nativeAd?.delegate = adListener
        return nativeAd
    }
    
    internal final class Listener: NSObject, APDNativeAdQueueDelegate, APDNativeAdPresentationDelegate {
        
        private let adChannel: FlutterMethodChannel
        
        fileprivate init(adChannel: FlutterMethodChannel) {
            self.adChannel = adChannel
        }
        
        func adQueueAdIsAvailable(_ adQueue: APDNativeAdQueue, ofCount count: UInt) {
            adChannel.invokeMethod("onNativeLoaded", arguments: nil)
            print("The value is \(adQueue.currentAdCount)")
        }
        
        func adQueue(_ adQueue: APDNativeAdQueue, failedWithError error: Error) {
            adChannel.invokeMethod("onNativeFailedToLoad", arguments: nil)
        }
        
        func nativeAdWillLogImpression(_ nativeAd: APDNativeAd) {
            adChannel.invokeMethod("onNativeShown", arguments: nil)
        }
        
        func nativeAdWillLogUserInteraction(_ nativeAd: APDNativeAd) {
            adChannel.invokeMethod("onNativeClicked", arguments: nil)
        }
        
        func nativeAdDidExpired(_ nativeAd: APDNativeAd) {
            adChannel.invokeMethod("onNativeExpired", arguments: nil)
        }
    }
}
